### PR TITLE
gh-96821: Add config option `--with-strict-overflow`

### PR DESCRIPTION
**Pull request by matthiasgoergens**
Please see the issue for details.

This PR marks the modules that need `-fno-strict-overflow` and adds machinery to force `-fstrict-overflow`, if the users wants that.

As far as is currently known, the three remaining modules that rely on defined integer overflow are fixed by:

* `_struct`: https://github.com/python/cpython/pull/96739
* `audioop`: https://github.com/python/cpython/pull/96923
* `_ctypes`: https://github.com/python/cpython/pull/96925

* Issue: gh-96821

### CODE REVIEW

**Comment by matthiasgoergens** [comment]
@ericsnowcurrently Please have a look.

I think perhaps we should only add the flag when we detect that the compiler supports it? What do you think?

**Comment by erlend-aasland** [comment]
> I think perhaps we should only add the flag when we detect that the compiler supports it? What do you think?

Yes, please add a compiler flag check. FWIW, the autoconf changes themselves look good.

File: Misc/NEWS.d/next/Build/2022-09-14-10-38-15.gh-issue-96821.Zk2a9c.rst
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Build/2022-09-14-10-38-15.gh-issue-96821.Zk2a9c.rst]: I'm not sure if the parenthesised sentence makes sense as part of a NEWS entry. Just describing the actual change should suffice.

Comment (matthiasgoergens) [reply Misc/NEWS.d/next/Build/2022-09-14-10-38-15.gh-issue-96821.Zk2a9c.rst]: I wanted to give a bit of context. But I'm happy to change it.

**Review by ericsnowcurrently — COMMENTED** [review]
This seems okay to me too.

First, though:

How can we verify (now or in the future) that these modules can't use strict overflow or that everything else can?

**Comment by ericsnowcurrently** [comment]
> How can we verify (now or in the future) that these modules can't use strict overflow or that everything else can?

Let's discuss this over on [gh-96821](https://github.com/python/cpython/issues/96821#issuecomment-1249667554).

File: Misc/NEWS.d/next/Build/2022-09-14-10-38-15.gh-issue-96821.Zk2a9c.rst
Comment (matthiasgoergens) [reply Misc/NEWS.d/next/Build/2022-09-14-10-38-15.gh-issue-96821.Zk2a9c.rst]: Please see new version.

File: Misc/NEWS.d/next/Build/2022-09-14-10-38-15.gh-issue-96821.Zk2a9c.rst
Comment (CAM-Gerlach) [inline Misc/NEWS.d/next/Build/2022-09-14-10-38-15.gh-issue-96821.Zk2a9c.rst]:
```suggestion
Explicitly mark C modules that need defined signed integer overflow,
and set up machinery for ``-fstrict-overflow``.
```

I assume you want to use a literal here, not the default role (italic prose)

File: configure.ac
Comment (erlend-aasland) [inline configure.ac]: Please use `dnl` instead of `#` for comments.

File: configure.ac
Comment (erlend-aasland) [inline configure.ac]: autoconf-good-practice-nit: please wrap all macro arguments in braces :)

Comment (erlend-aasland) [inline configure.ac]: autoconf-good-practice-nit: please wrap all macro arguments in braces :)

Also, I think brevity is better: _"default is no for backwards compatibility"_ => _"default is no"_

File: Misc/NEWS.d/next/Build/2022-09-14-10-38-15.gh-issue-96821.Zk2a9c.rst
Comment (erlend-aasland) [reply Misc/NEWS.d/next/Build/2022-09-14-10-38-15.gh-issue-96821.Zk2a9c.rst]: My personal preference would be to name them _C **extension** modules_. Also, I think "machinery" is too vague. Why not just describe what this PR does?

File: configure.ac
Comment (matthiasgoergens) [reply configure.ac]: I just copied the style I found in the rest of the file.

By braces, do you mean `{}`?  From what I can find in the docs, single or double square brackets might be the way to go?

File: Misc/NEWS.d/next/Build/2022-09-14-10-38-15.gh-issue-96821.Zk2a9c.rst
Comment (matthiasgoergens) [reply Misc/NEWS.d/next/Build/2022-09-14-10-38-15.gh-issue-96821.Zk2a9c.rst]: I'm not sure they are all extension modules?  Basically, I marked anything that's written in C that needs `-fno-struct-overflow`, no matter whether it's some extension or the core of the eval loop.

(The latter one has since been fixed, so that's no longer in this PR.  But it would have been marked.)

In any case, I implemented your suggestion.

File: configure.ac
Comment (erlend-aasland) [inline configure.ac]:
```suggestion
AC_MSG_CHECKING([for --with-strict-overflow])
AC_ARG_WITH([strict-overflow],
```

File: configure.ac
Comment (erlend-aasland) [inline configure.ac]:
```suggestion
    [choose between -fstrict-overflow or -fno-strict-overflow (default is no)]),[],
    [with_strict_overflow=no])
```

File: configure.ac
Comment (erlend-aasland) [inline configure.ac]: Perhaps something like this?

```suggestion
dnl is defined behaviour via twos-complement.
dnl Set STRICT_OVERFLOW_CFLAGS and NO_STRICT_OVERFLOW_CFLAGS depending on compiler support.
dnl Pass the latter to modules that depend on such behaviour.
```

**Comment by matthiasgoergens** [comment]
@erlend-aasland All excellent suggestions!

**Review by erlend-aasland — APPROVED** [review]
lgtm

**Comment by matthiasgoergens** [comment]
Let me remove the special handling for `_struct`, because that PR is already merged. And then we can merge this one?

**Comment by matthiasgoergens** [comment]
OK, this is ready to merge as far as I am concerned. @erlend-aasland

**Comment by erlend-aasland** [comment]
I'll wait for a final thumbs-up from @ericsnowcurrently before merging.

**Comment by ericsnowcurrently** [comment]
This seems good to me.  However, I don't have the expertise to review the configure script changes.

**Comment by erlend-aasland** [comment]
@matthiasgoergens, can you please document the new configure option and add a link to those docs in the NEWS entry? See `Doc/using/configure.rst` and the `:option:` Sphinx directive.

**Comment by matthiasgoergens** [comment]
> @matthiasgoergens, can you please document the new configure option and add a link to those docs in the NEWS entry? See `Doc/using/configure.rst` and the `:option:` Sphinx directive.

Done!

File: Doc/using/configure.rst
Comment (AA-Turner) [inline Doc/using/configure.rst]: This should fix docs CI:

```suggestion
   Add ``-fstrict-overflow`` to the C compiler flags (by default we add
   ``-fno-strict-overflow`` instead).
```

File: Misc/NEWS.d/next/Build/2022-09-14-10-38-15.gh-issue-96821.Zk2a9c.rst
Comment (AA-Turner) [inline Misc/NEWS.d/next/Build/2022-09-14-10-38-15.gh-issue-96821.Zk2a9c.rst]: Per @erlend-aasland 
```suggestion
and add a configure option :option:`--with-strict-overflow`.
```

**Comment by matthiasgoergens** [comment]
@AA-Turner Thanks for the suggestions!  Much appreciated.

**Review by erlend-aasland — DISMISSED** [review]

File: configure.ac
Comment (erlend-aasland) [inline configure.ac]: Sorry to pick on this further. I would like two more changes before merging:

1. The help message is too vague. I'd like something a la: _if 'yes', add -fstrict-overflow to CFLAGS, else add -fno-strict-overflow to CFLAGS (default is no)_.
2. Please add a `AC_MSG_WARN` if `ac_cv_cc_supports_fstrict_overflow=no` and `--with-strict-overflow=yes` (_"--with-strict-overflow=yes specified, but your compiler does not support the -fstrict-overflow flag"_

Comment (matthiasgoergens) [reply configure.ac]: Makes sense. Don't worry about picking on this, I'm just glad you care enough to make the effort!

Comment (erlend-aasland) [reply configure.ac]: Thanks, just ping me when you've made the changes, and I'll get this merged.

**Comment by ericsnowcurrently** [comment]
Thanks for shepherding this through, @erlend-aasland!

**Comment by smontanaro** [comment]
(I'm working my way through some PRs which have been approved and are labeled "awaiting merge", hence my seemingly bolt from the blue comment. Why? [Read here](https://discuss.python.org/t/prs-awaiting-merge/21313).)

This work looks to be completed. Can someone merge?

**Comment by erlend-aasland** [comment]
> (I'm working my way through some PRs which have been approved and are labeled "awaiting merge", hence my seemingly bolt from the blue comment. Why? [Read here](https://discuss.python.org/t/prs-awaiting-merge/21313).)
> 
> This work looks to be completed. Can someone merge?

I'm awaiting final changes. See https://github.com/python/cpython/pull/96823#discussion_r997896383

**Comment by hauntsaninja** [comment]
@erlend-aasland I was interested in picking this up, so I added the help message change you suggested here.

I was looking into adding the warn as well, but I'm not familiar with autoconf. Should we be setting `ac_cv_cc_supports_fstrict_overflow` in the branches of the `AC_COMPILE_IFELSE`?

**Comment by matthiasgoergens** [comment]
Please pick this up, if you can. Thanks!

I don't have as much time as I would like these days.

**Comment by hauntsaninja** [comment]
Okay, I think this is ready. I'm not familiar with autoconf, so apologies for any issues!

**Comment by erlend-aasland** [comment]
Thanks, Shantanu! It looks good, AFAICS. As an improvement, we could rewrite the logic to only check for the compiler switch if `--with-strict-overflow` was given.

**Comment by CAM-Gerlach** [comment]
Gentle ping @hauntsaninja @erlend-aasland This came up in discussion and it looks like we're really close to being `-fstrict-overflow` clean here. Anything else this needs to be merged for 3.12, besides reviewing and merging #96925 also? From a (minimal) docs review perspective, this looks good. Thanks!

**Comment by hauntsaninja** [comment]
I think this is ready to merge if we need it. Erlend's last comment is a nice to have and I can address it in a follow up PR.

**Comment by matthiasgoergens** [comment]
Nice to see this progressing!

**Comment by erlend-aasland** [comment]
Thanks Shantanu and Matthias!

**Pull request merged**
